### PR TITLE
feat: new dataset/table/column models

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

This PR implements new models for SIP-68 (https://github.com/apache/superset/issues/14909):

- `Column`
- `Table`
- `Dataset`

The associated tables are created with a migration, and the existing datasets (`SqlaTable`, `SqlMetric`, `TableColumn`) are migrated to the new models:

- A physical dataset is mapped into a `Table` instance and a `Dataset` instance.
- A virtual dataset is mapped into a `Dataset` instance.
- Metrics and columns are mapped into `Column` instances (in the new model it represents physical columns, derived columns, and metrics).

The models are kept up-to-date via hooks added to `SqlaTable`, `SqlMetric`, and `TableColumn`. Every time these models are created, deleted, or updated, the new models are updated correspondingly. The sync is unidirectional, from the old models to the new ones.

The next step is to modify the backend to **read** from the new models. Initially the API will remain unmodified, but hopefully in the future we might be able to clean it up and modernize it (in some places we haven't migrated to the v1 API yet).

After that, the backend will be modified to **write** to the new models. Once the backend is reading and writing from the new models we can get rid of the old models (including the Druid models).

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

No visual changes.

### TESTING INSTRUCTIONS

Reviewers can run `superset load-examples` and see the new models being populated:

```sql
mysql> SELECT `schema`, name FROM sl_tables;
+----------+------------------------+
| schema   | name                   |
+----------+------------------------+
| examples | wb_health_population   |
| examples | birth_names            |
| examples | long_lat               |
| examples | birth_france_by_region |
| examples | sf_population_polygons |
| examples | flights                |
| examples | bart_lines             |
| examples | channel_members        |
| examples | messages               |
| examples | channels               |
| examples | threads                |
| examples | unicode_test           |
| examples | users                  |
| examples | users_channels         |
| examples | cleaned_sales_data     |
+----------+------------------------+
15 rows in set (0.00 sec)

mysql> SELECT name, is_physical FROM sl_datasets;
+---------------------------+-------------+
| name                      | is_physical |
+---------------------------+-------------+
| wb_health_population      |           1 |
| birth_names               |           1 |
| long_lat                  |           1 |
| birth_france_by_region    |           1 |
| sf_population_polygons    |           1 |
| flights                   |           1 |
| bart_lines                |           1 |
| FCC 2018 Survey           |           0 |
| messages_channels         |           0 |
| channel_members           |           1 |
| covid_vaccines            |           0 |
| messages                  |           1 |
| channels                  |           1 |
| members_channels_2        |           0 |
| exported_stats            |           0 |
| threads                   |           1 |
| new_members_daily         |           0 |
| unicode_test              |           1 |
| users                     |           1 |
| video_game_sales          |           0 |
| users_channels            |           1 |
| cleaned_sales_data        |           1 |
| users_channels-uzooNNtSRO |           0 |
+---------------------------+-------------+
23 rows in set (0.00 sec)

mysql> SELECT expression FROM sl_columns WHERE is_aggregation;
+-----------------------+
| expression            |
+-----------------------+
| sum(`SP_POP_TOTL`)    |
| sum(`SH_DYN_AIDS`)    |
| sum(`SP_RUR_TOTL_ZS`) |
| sum(`SP_DYN_LE00_IN`) |
| sum(`SP_RUR_TOTL`)    |
| COUNT(*)              |
| COUNT(*)              |
| SUM(num)              |
| COUNT(*)              |
| AVG(`2004`)           |
| COUNT(*)              |
| COUNT(*)              |
| COUNT(*)              |
| COUNT(*)              |
| count(*)              |
+-----------------------+
15 rows in set (0.02 sec)
```

The PR has unit tests checking that the sync works on creation, deletion, and updates.

### ADDITIONAL INFORMATION

- [X] Has associated issue: https://github.com/apache/superset/issues/14909
- [ ] Required feature flags:
- [ ] Changes UI
- [X] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [X] Migration is atomic, supports rollback & is backwards-compatible
  - [X] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -1790,23 +1806,448 @@ def before_update(
...
+        mapper: Mapper, connection: Connection, target: Union[SqlMetric, TableColumn]
     ) -> None:
         """
         Forces an update to the table's changed_on value when a metric or column on the
         table is updated. This busts the cache key for all charts that use the table.
 
-        :param _mapper: Unused.
-        :param _connection: Unused.
-        :param obj: The metric or column that was updated.
+        :param mapper: Unused.
```
Comment (craig-rueda) [inline superset/connectors/sqla/models.py:1816]: Would it make sense to pull this hook logic out into each "New" model's respective classes? This is a rather large block and could use a little breakup :).

File: superset/connectors/sqla/models.py [hunk 2]
```diff
@@ -71,12 +72,14 @@
...
 
 from superset import app, db, is_feature_enabled, security_manager
+from superset.columns.models import Column as NewColumn
 from superset.common.db_query_status import QueryStatus
 from superset.connectors.base.models import BaseColumn, BaseDatasource, BaseMetric
 from superset.connectors.sqla.utils import (
     get_physical_table_metadata,
     get_virtual_table_metadata,
 )
+from superset.datasets.models import Dataset as NewDataset
```
Comment (craig-rueda) [inline superset/connectors/sqla/models.py:82]: Maybe alias this as something like `DatasetV2`? I get that it's "new and shiny" but it won't always be the case :).

File: superset/migrations/versions/b8d3a24d9131_new_dataset_models.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+"""
+
+from uuid import uuid4
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy_utils import UUIDType
+
+from superset import db
+from superset.connectors.sqla.models import SqlaTable
```
Comment (craig-rueda) [inline superset/migrations/versions/b8d3a24d9131_new_dataset_models.py:32]: Don't import models directly in migrations. If (when) we ever update this model, the migrations might (will) break.

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (betodealmeida) [reply superset/connectors/sqla/models.py:1816]: I left it in the old model because it will go away with the old model — this is only needed while both exist. But I agree it's hard to read, I'll break it down. There's also some duplicate logic, but since this is temporary code that's going away I thought it would be ok.

File: superset/connectors/sqla/models.py (hunk 2 above)
Comment (betodealmeida) [reply superset/connectors/sqla/models.py:82]: After this PR the plan is to get rid of the old models, so this will go away before we have v3.

File: superset/migrations/versions/b8d3a24d9131_new_dataset_models.py (hunk 3 above)
Comment (betodealmeida) [reply superset/migrations/versions/b8d3a24d9131_new_dataset_models.py:32]: Ah, great point. Let me fix it.

File: superset/tables/models.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class Table(Model, AuditMixinNullable, ExtraJSONMixin, ImportExportMixin):
+    """
+    A table/view in a database.
+    """
+
+    __tablename__ = "sl_tables"
+
+    # Note this uniqueness constraint is not part of the physical schema, i.e., it does
```
Comment (john-bodley) [inline superset/tables/models.py:57]: Thanks for including this comment. I know it's something which has tripped us up in the past.

File: tests/unit_tests/tables/test_models.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        catalog="my_catalog",
+        database=Database(database_name="my_database", sqlalchemy_uri="test://"),
+        columns=[Column(name="ds", type="TIMESTAMP", expression="ds",)],
+    )
+    session.add(table)
+    session.flush()
+
+    assert table.id == 1
+    assert table.uuid is not None
+
```
Comment (john-bodley) [inline tests/unit_tests/tables/test_models.py:46]:
```suggestion
```

File: tests/unit_tests/tables/test_models.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    assert table.database_id == 1
+
```
Comment (john-bodley) [inline tests/unit_tests/tables/test_models.py:48]:
```suggestion
```

File: tests/unit_tests/tables/test_models.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    assert table.catalog == "my_catalog"
+    assert table.schema == "my_schema"
+    assert table.name == "my_table"
+
```
Comment (john-bodley) [inline tests/unit_tests/tables/test_models.py:52]:
```suggestion
```

**Review by john-bodley — COMMENTED** [review]
Thanks @betodealmeida for making this change—and adding a slew of unit tests to boot. I think your dual write approach is the right approach and allows us (the community) to iron out any kinks in the system before making the switch.

File: superset/columns/models.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    # We use ``sa.Text`` for these attributes because (1) in modern databases the
+    # performance is the same as ``VARCHAR``[1] and (2) because some table names can be
+    # **really** long (eg, Google Sheets URLs).
+    #
+    # [1] https://www.postgresql.org/docs/9.1/datatype-character.html
+    name = sa.Column(sa.Text)
+    type = sa.Column(sa.Text)
+
+    # Columns are defined by expressions. For tables, these are the actual columns names,
```
Comment (john-bodley) [inline superset/columns/models.py:60]: This is the one thing I find somewhat atypical, i.e., that a physical table/view column would require require and expression which needs to match the name. Why not just make `expression` nullable in that case and thus the column would only represent actual SQL expressions.

File: superset/columns/models.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+):
+    """
+    A "column".
+
+    The definition of column here is overloaded: it can represent a physical column in a
+    database relation (table or view); a computed/derived column on a dataset; or an
+    aggregation expression representing a metric.
+    """
+
+    __tablename__ = "sl_columns"
```
[the remark this answers is not in the document]
Comment (betodealmeida) [reply superset/columns/models.py:48]: Semantic layer. I wanted to use `columns` and `tables`, but those are taken by `TableColumn` and `SqlaTable`. Originally I had the migration rename these tables, but it would force us to have downtime when upgrading, so I opted for using a prefix instead.

I'm open to other suggestions for a prefix here. :)

File: superset/columns/models.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    is_aggregation = sa.Column(sa.Boolean, default=False)
+
+    # Assuming the column is an aggregation, is it additive? Useful for determining which
+    # aggregations can be done on the metric. Eg, ``COUNT(DISTINCT user_id)`` is not
+    # additive, so it shouldn't be used in a ``SUM``.
+    is_additive = sa.Column(sa.Boolean, default=False)
+
+    # Is an increase good? Useful for displaying the results of A/B tests, or setting up
+    # alerts. Eg, this is true for "revenue", but false for "latency".
+    increase_good = sa.Column(sa.Boolean, default=True)
```
[the remark this answers is not in the document]
Comment (betodealmeida) [reply superset/columns/models.py:95]: Ah, good point, I'll change it.

File: superset/columns/models.py (hunk 8 above)
Comment (betodealmeida) [reply superset/columns/models.py:60]: I'm hoping that having `expression` even for physical columns and tables will make things easier because of the consistency. One thing the new model does is automatically quote table/column names that are also identifiers; as an example, we'd have:

```python
column = Column(
    name="select",
    expression="`select`",  # or "select", depending on the DB
    ...
)
```

So to select that column we just need to use its `expression`. Otherwise, we'd have to check if `expression` is null, and then potentially encode the name in order to select that column.

File: superset/columns/schemas.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+These models are not fully implemented, and shouldn't be used yet.
+"""
+
+from marshmallow_sqlalchemy import SQLAlchemyAutoSchema
+
+from superset.columns.models import Column
+
+
+class ColumnSchema(SQLAlchemyAutoSchema):
```
Comment (dpgaspar) [inline superset/columns/schemas.py:32]: Where is this used? are you planning to include it on a followup PR?
It's a fast and dynamic way of generating Marshmallow schemas but can lack some custom validations and will not have descriptions for example, those are a nice to have on OpenAPI spec

Comment (betodealmeida) [reply superset/columns/schemas.py:32]: This is used in this PR just for unit tests, I'll expand it into a manual schema later when we start reading from these models.

File: superset/columns/models.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Column model.
+
+This model was introduced in SIP-68 (https://github.com/apache/superset/issues/14909),
+and represents a "column" in a table or dataset. In addition to a column, new models for
+tables, metrics, and datasets were also introduced.
```
Comment (eschutho) [inline superset/columns/models.py:22]: Other than the connectors, don't we normally put the model files in the superset/models/* folders?

File: superset/columns/models.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    A "column".
+
+    The definition of column here is overloaded: it can represent a physical column in a
+    database relation (table or view); a computed/derived column on a dataset; or an
+    aggregation expression representing a metric.
+    """
+
+    __tablename__ = "sl_columns"
+
+    id = sa.Column(sa.Integer, primary_key=True)
```
Comment (eschutho) [inline superset/columns/models.py:50]: not a blocker, but more of a question.. when do we want to start using uuids instead of auto-incrementing integers for  public-facing ids?

File: superset/columns/models.py (hunk 12 above)
Comment (betodealmeida) [reply superset/columns/models.py:22]: I tried to have this aligned with https://github.com/apache/superset/issues/9077. `superset/models/` came before that (I think). But I have no strong preference here.

Comment (betodealmeida) [reply superset/columns/models.py:22]: @dpgaspar @john-bodley  any thoughts on this?

File: superset/columns/models.py (hunk 13 above)
Comment (betodealmeida) [reply superset/columns/models.py:50]: I think we would need a SIP and a major release for that, no?

File: tests/unit_tests/columns/__init__.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (pkdotson) [inline tests/unit_tests/columns/__init__.py:1]: is this file necessary?

Comment (betodealmeida) [reply tests/unit_tests/columns/__init__.py:1]: It is... in theory in Python 3 empty `__init__py` files are not needed, but in practice we had a lot of problems when we omit them.

For `pytest` in particular, if you have files with the same name in different directories inside your tests, like:

- `tests/unit_tests/datasets/test_model.py`
- `tests/unit_tests/tables/test_model.py`

Then you *need* to add the `__init__.py` file to the subdirectories, otherwise pytest will only read one of them.

File: superset/columns/models.py (hunk 12 above)
Comment (eschutho) [reply superset/columns/models.py:22]: It looks like @willbarrett has models at the top level in that Sip, but I don't have much more context than that.

File: superset/columns/models.py (hunk 13 above)
Comment (eschutho) [reply superset/columns/models.py:50]: Since this is a new model, I don't see any reason why we couldn't add it in here since it wouldn't break anything, but yes I agree on a SIP regarding an overall change toward that methodology. If that's the blocker, then that makes sense.

**Review by hughhhh — APPROVED** [review]

File: superset/columns/models.py (hunk 13 above)
Comment (betodealmeida) [reply superset/columns/models.py:50]: Ah, I see your point. But the idea is that at least initially the APIs would be unchanged, we'd just reroute the backend to read from the new models instead of the old ones.

**Comment by villebro** [comment]
This is impressive work @betodealmeida ! 🚀 Talk about swapping out the airplane's engines mid flight! ✈️

File: superset/tables/models.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    # not exist in the migrations. The reason it does not physically exist is MySQL,
+    # PostgreSQL, etc. have a different interpretation of uniqueness when it comes to NULL
+    # which is problematic given the catalog and schema are optional.
+    __table_args__ = (UniqueConstraint("database_id", "catalog", "schema", "name"),)
```
Comment (ktmud) [inline superset/tables/models.py:61]: @betodealmeida It's also not possible to apply this constraint in MySQL because TEXT cannot be used in index keys without specifying a length: https://stackoverflow.com/questions/1827063/mysql-error-key-specification-without-a-key-length
